org/mozilla/jss/ssl/javax/BufferPRFD.h

```diff
@@ -0,0 +1,344 @@
+#include <string.h>
+
+// Use a modern pragma guard...
+#pragma once
+
+/* This struct stores all the private data we need access to from inside our
+ * PRFileDesc calls. We store the following information:
+ *
+ *      read_bytes -- raw buffer to read from for recv(...) calls
+ *      read_capacity -- size of read_bytes buffer
+ *      read_ptr -- location after the last character in the buffer;
+ *                  i.e., location to place the next character if writing.
+ *
+ *      write_bytes -- raw buffer to write to for send(...) calls
+ *      write_capacity -- size of write_bytes buffer
+ *      write_ptr -- location after the last character in the buffer;
+ *                   i.e., location to place the next character if writing.
+ *
+ *      closed -- whether or not the FileDesc is closed.
+ *
+ * As the read_* and write_* members should be provided by the creator, and
+ * actors outside our PRFileDesc need access to this information (to add more
+ * bytes to the read buffer when more data arrives for instance), we store
+ * pointers and not the values themselves.
+ *
+ * The creator is responsible for ensuring that all data gets correctly freed
+ * when the program exits; we will not free any of our pointers that were not
+ * created by us.
+ */
+struct PRFilePrivate {
+    uint8_t* read_bytes;
+    size_t* read_capacity;
+    size_t* read_ptr;
+
+    uint8_t* write_bytes;
+    size_t* write_capacity;
+    size_t* write_ptr;
+
+    uint8_t* peer_addr;
+};
+
+// This function is provided as a stub for all unimplemented calls.
+static PRIntn invalidInternalCall(/* anything */)
+{
+    /* For debugging; any internal calls are asserted, so a backtrace is
+     * displayed so _hopefully_ we get information about which call was
+     * performed. To enable asserts, define DEBUG or FORCE_PR_ASSERT before
+     * loading NSPR headers.
+     */
+    PR_ASSERT(!"invalidInternalCall performed!");
+    return 0;
+}
+
+// This function mimics shutting down a buffer.
+static PRStatus PRBufferShutdown(PRFileDesc *fd, PRShutdownHow how)
+{
+    // This method is currently only a stub with no functionality, as
+    // we'd have to track buffer state and the SSL engine might still
+    // be waiting on us to unblock before writing more data.
+    return PR_SUCCESS;
+}
+
+// This function mimics closing a buffer and frees our associated data.
+static PRStatus PRBufferClose(PRFileDesc* fd)
+{
+    // This method is the last call our buffer will see, so free our data.
+    if (fd == NULL) {
+        // So we don't dereference a NULL pointer in case we've been closed
+        // multiple times, exit with success.
+        return PR_SUCCESS;
+    }
+
+    if (fd->secret != NULL) {
+        /* In case the data on the wire is sensitive (it shouldn't be -- all
+         * incoming and outgoing data should be encrypted except for the
+         * client hello -- memset the private structure first. */
+        memset(fd->secret, 0, sizeof(PRFilePrivate));
```

Comment:
This comment and `memset` call are actually out of date -- the structure itself only contains pointers to data and not the data itself. I think the `memset` call is extraneous and only ensures that our internal pointers are null in case anyone maintains a pointer to it.

We _can't_ actually free or zero the internal pointers as we don't own them and there may still be data (e.g., a `CLOSE_NOTIFY` event) that needs to be sent.